### PR TITLE
[5.x] S3 collection spec parsing

### PR DESCRIPTION
**Pull request by tdrwenski**
- S3 version of the collection spec parser with an abstract base class
- Tests for the S3 collection spec parsing
- Fix MFileCollectionManager for S3 collection specs
- Add tests for MFileCollectionManager, local and s3, to verify files given by spec are being correctly found

### CODE REVIEW

File: cdm/core/src/test/java/thredds/inventory/TestCollectionSpecParser.java [hunk 1]
```diff
@@ -58,10 +58,10 @@ public static List<Object[]> getTestParameters() throws URISyntaxException {
         {"/data/ldm/pub/decoded/netcdf/surface/metar/**/", "/data/ldm/pub/decoded/netcdf/surface/metar", true, null,
             null},
 
-        {"/data/ldm/pub/decoded/netcdf/surface/metar/**/*", "/data/ldm/pub/decoded/netcdf/surface/metar", true, null,
+        {"/data/ldm/pub/decoded/netcdf/surface/metar/**/.*", "/data/ldm/pub/decoded/netcdf/surface/metar", true, ".*",
```
Comment (tdrwenski) [inline cdm/core/src/test/java/thredds/inventory/TestCollectionSpecParser.java]: I very slightly changed the behavior of the original `CollectionSpecParser`. Previously, if the regEx was only one character (here '\*' but could be any character) then it resulted in `filter = null`. Now, a regex of '*' will be compiled into a `Pattern` and will fail.

Comment (haileyajohnson) [reply cdm/core/src/test/java/thredds/inventory/TestCollectionSpecParser.java]: hmm, I'm a little worried that will break some users, I think the single '*' pattern is pretty commonly used...

Comment (tdrwenski) [reply cdm/core/src/test/java/thredds/inventory/TestCollectionSpecParser.java]: It just sort of seemed like a mistake that a one character filter was ignored. You can achieve the same results with no filter or with ".*".

But regardless that could be annoying for users already using a "*" so I can change it back.

Comment (haileyajohnson) [reply cdm/core/src/test/java/thredds/inventory/TestCollectionSpecParser.java]: yeah, I think '**/*' is pretty standard for "anything at any nesting level", even if it is unnecessary

Comment (tdrwenski) [reply cdm/core/src/test/java/thredds/inventory/TestCollectionSpecParser.java]: back to original behavior in 35143be7db

**Review by haileyajohnson — APPROVED** [review]

**Pull request merged**